### DISCUSSION TITLE
Example to build a custom renderer

### DISCUSSION THREAD

**Discussion by flexdinesh** [markdoc/markdoc#288] [Q&A] [answered]
I'm looking to use Markdoc with Astro. I'm happy to see that [Markdoc supports building custom renderers](https://markdoc.dev/docs/render#create-your-own). Is there an example somewhere that I could use as an inspiration to build a custom renderer?

**Comment by rpaul-stripe** [comment, marked as answer]
The [HTML and React renderers](https://github.com/markdoc/markdoc/tree/main/src/renderers) that come with Markdoc are probably the best starting point. I also recommend outputting the render tree from some sample Markdoc content and taking a close look at the structure. The render tree mostly consists of [tag instances](https://github.com/markdoc/markdoc/blob/main/src/tag.ts), values, and arrays for fragments. A custom renderer just has to convert the render tree data structure into the desired output format.

**Reply by flexdinesh** [reply]
Thanks @rpaul-stripe. I built a renderer for Astro and it works. Everything is local now but I'll soon open source the code and share it here in case someone else needs to build a rendered for Astro.

**Reply by mfix-stripe** [reply]
There is also [this example for Vue](https://github.com/wobsoriano/vue-markdoc/blob/master/src/index.ts) 🙂

**Reply by abirmingham** [reply]
Hopping on this thread, as I am currently writing a renderer as well. In my case, I need to produce plain-old-markdown from Markdoc. One initial question - `Tag.name` is typed as the generic `N extends string = string`. Is there a list somewhere of all the builtin tag names that Markdoc produces for a renderable tree node? Currently I am reverse engineering the list examining the renderable tree produced by my own markdown, but I want to make sure my renderer handles the complete set.

Thank you!

This is my current list:
```
type TagType =
| 'a'
| 'article'
| 'blockquote'
| 'code'
| 'em'
| 'h1'
| 'h2'
| 'hr'
| 'img'
| 'li'
| 'p'
| 'pre'
| 'strong'
| 'ul';
```

**Reply by flexdinesh** [reply]
Forgot to share it here. I built an Astro renderer for markdoc and shipped it a few months ago.

https://www.npmjs.com/package/astro-markdoc-renderer

**Reply by mfix-stripe** [reply]
@flexdinesh I think that Astro renderer deserves its own "show and tell" Discussions post 🙂
